roachtest: improve multi-region TPC-C tests

otan opened the pull request:
This commit changes the multi-region roachtests to target survivability
better, whilst also outputting better metrics. In particular:

* Region survivability now tests shutting down an entire region.
  Workloads now only target one region at a time, with a test for
  cross-region queries.
* Zone survivability now tests shut down one node at a time.
* Added prometheus metrics which can be observed by the snapshot or
  during test invocation.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/66755)
<!-- Reviewable:end -->

otan commented:
Region survivability test:

![image](https://user-images.githubusercontent.com/3646147/123049532-ff3c3d00-d442-11eb-9e88-8e43aaa4703b.png)

Zone survivability test:
![image](https://user-images.githubusercontent.com/3646147/123049576-0b27ff00-d443-11eb-8e6a-2bb0b7551261.png)

ajstorm reviewed (changes requested):
Pretty cool.  We might want to have @nvanbenschoten have a look as well, as he's much more familiar with this test than I am.

> Reviewed 3 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66755#-:-McyH3qV1lm8M_JP5ddE:bd9g8rq)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/cmd/roachtest/tpcc.go, line 72 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66755#-McyB--M3JCf6VFp6dzD:-McyB--M3JCf6VFp6dzE:ba4cp6z) ([raw file](https://github.com/cockroachdb/cockroach/blob/d663594e5875825313325bf7836e81d40b526b95/pkg/cmd/roachtest/tpcc.go#L72)):*
> ```Go
> 
> type workloadInstance struct {
> 	// nodes dictates the workload should run against.
> ```

Nit "dictates the nodes the workload should run against?"
___
*[pkg/cmd/roachtest/tpcc.go, line 219 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66755#-McyBdN1DP6hrOU4TzuH:-McyBdN1DP6hrOU4TzuI:bwandht) ([raw file](https://github.com/cockroachdb/cockroach/blob/d663594e5875825313325bf7836e81d40b526b95/pkg/cmd/roachtest/tpcc.go#L219)):*
> ```Go
> 	m := newMonitor(ctx, c, crdbNodes)
> 	for i := range workloadInstances {
> 		i := i
> ```

Nit: Is this required to create a copy of i to send to the go routine below?  If so, you might want to either add a comment, or change the name of the local copy of this variable so that someone doesn't come along in the future and delete it.
___
*[pkg/cmd/roachtest/tpcc.go, line 221 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66755#-McyC5KG8mJhFHkCytmx:-McyC5KG8mJhFHkCytmy:b-t7qsql) ([raw file](https://github.com/cockroachdb/cockroach/blob/d663594e5875825313325bf7836e81d40b526b95/pkg/cmd/roachtest/tpcc.go#L221)):*
> ```Go
> 		i := i
> 		m.Go(func(ctx context.Context) error {
> 			t.WorkerStatus("running tpcc")
> ```

Nit: Does it make sense to expand this status to indicate which workload instance this is?
___
*[pkg/cmd/roachtest/tpcc.go, line 438 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66755#-McyCw5T267UPmUGBwYy:-McyCw5T267UPmUGBwYz:bf50l35) ([raw file](https://github.com/cockroachdb/cockroach/blob/d663594e5875825313325bf7836e81d40b526b95/pkg/cmd/roachtest/tpcc.go#L438)):*
> ```Go
> 	// Setup multi-region tests.
> 	{
> 		zs := []string{
> ```

Nit: I know you didn't change this, but it seems error-prone to me that these two slices depend on each other (as well as on the total number of nodes) and yet we're not validating that anywhere.  Can we pull up the total number of nodes into a constant, and validate that there are enough zones and nodes for the defined regions? 

Also, shouldn't the number of zones per region match the number of nodes per region?  Otherwise we're going to have 3 nodes in each region but all of those nodes will be in the same AZ?
___
*[pkg/cmd/roachtest/tpcc.go, line 459 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66755#-McyEpWl5nrJ6H5r452h:-McyEpWl5nrJ6H5r452i:b-x01ipd) ([raw file](https://github.com/cockroachdb/cockroach/blob/d663594e5875825313325bf7836e81d40b526b95/pkg/cmd/roachtest/tpcc.go#L459)):*
> ```Go
> 				survivalGoal: "zone",
> 				chaosTarget: func(iter int) option.NodeListOption {
> 					return option.NodeListOption{(iter % (len(regions) * nodesPerRegion)) + 1}
> ```

I might be missing something, but it seems like we're computing the number of nodes to be dependant on the number of regions here (good) but then hard-coding the number of workload instances below to be 9 (less good).  Shouldn't the number of workload instances be proportional to the number of nodes/regions?
___
*[pkg/cmd/roachtest/tpcc.go, line 522 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66755#-McyFjWx3onoMjZDUUlQ:-McyFjWx3onoMjZDUUlR:b-42kehv) ([raw file](https://github.com/cockroachdb/cockroach/blob/d663594e5875825313325bf7836e81d40b526b95/pkg/cmd/roachtest/tpcc.go#L522)):*
> ```Go
> 				},
> 				workloadInstances: []workloadInstance{
> 					// For 2110 to 2112, talk to a single region whose data
> ```

Same comment down here about hard-coding.
___
*[pkg/cmd/roachtest/tpcc.go, line 567 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66755#-McyFuJzAmmm_gaX43SS:-McyFuK-BXlCaS0_py1i:bk4a44m) ([raw file](https://github.com/cockroachdb/cockroach/blob/d663594e5875825313325bf7836e81d40b526b95/pkg/cmd/roachtest/tpcc.go#L567)):*
> ```Go
> 				MinVersion: "v21.1.0",
> 				// 3 nodes per region + 1 node for workload.
> 				Cluster: r.makeClusterSpec(10, spec.Geo(), spec.Zones(strings.Join(zs, ","))),
> ```

hard coded number of nodes
___
*[pkg/cmd/roachtest/tpcc.go, line 588 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66755#-McyGSCG3lNPUJ5uP-K3:-McyGSCG3lNPUJ5uP-K4:b-h61e3z) ([raw file](https://github.com/cockroachdb/cockroach/blob/d663594e5875825313325bf7836e81d40b526b95/pkg/cmd/roachtest/tpcc.go#L588)):*
> ```Go
> 					iter := 0
> 					// TODO(#multiregion): setup workload to run specifically for a given partition
> 					// on each node of a cluster, instead of one node using a workload on all clusters.
> ```

Is this the TODO we're addressing with this PR?


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66755)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm)

___
*[pkg/cmd/roachtest/tpcc.go, line 219 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66755#-McyBdN1DP6hrOU4TzuH:-Md-WDND8kOTu2IAtQdi:b-kl1mki) ([raw file](https://github.com/cockroachdb/cockroach/blob/d663594e5875825313325bf7836e81d40b526b95/pkg/cmd/roachtest/tpcc.go#L219)):*
<details><summary><i>Previously, ajstorm (Adam Storm) wrote…</i></summary><blockquote>

Nit: Is this required to create a copy of i to send to the go routine below?  If so, you might want to either add a comment, or change the name of the local copy of this variable so that someone doesn't come along in the future and delete it.
</blockquote></details>

This is a pretty standard go practice with that implication, but i added it anyway.
___
*[pkg/cmd/roachtest/tpcc.go, line 459 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66755#-McyEpWl5nrJ6H5r452h:-Md-_E-V12gZySl-vd6q:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d663594e5875825313325bf7836e81d40b526b95/pkg/cmd/roachtest/tpcc.go#L459)):*
<details><summary><i>Previously, ajstorm (Adam Storm) wrote…</i></summary><blockquote>

I might be missing something, but it seems like we're computing the number of nodes to be dependant on the number of regions here (good) but then hard-coding the number of workload instances below to be 9 (less good).  Shouldn't the number of workload instances be proportional to the number of nodes/regions?
</blockquote></details>

Done.
___
*[pkg/cmd/roachtest/tpcc.go, line 522 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66755#-McyFjWx3onoMjZDUUlQ:-Md-_EnbAEIdf0uNgtRe:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d663594e5875825313325bf7836e81d40b526b95/pkg/cmd/roachtest/tpcc.go#L522)):*
<details><summary><i>Previously, ajstorm (Adam Storm) wrote…</i></summary><blockquote>

Same comment down here about hard-coding.
</blockquote></details>

Done.
___
*[pkg/cmd/roachtest/tpcc.go, line 567 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66755#-McyFuJzAmmm_gaX43SS:-Md-_FXp7FNMQc43Pvvg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d663594e5875825313325bf7836e81d40b526b95/pkg/cmd/roachtest/tpcc.go#L567)):*
<details><summary><i>Previously, ajstorm (Adam Storm) wrote…</i></summary><blockquote>

hard coded number of nodes
</blockquote></details>

Done.
___
*[pkg/cmd/roachtest/tpcc.go, line 588 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66755#-McyGSCG3lNPUJ5uP-K3:-Md-TK3w3pRvxW8RKhed:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d663594e5875825313325bf7836e81d40b526b95/pkg/cmd/roachtest/tpcc.go#L588)):*
<details><summary><i>Previously, ajstorm (Adam Storm) wrote…</i></summary><blockquote>

Is this the TODO we're addressing with this PR?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

otan committed:
roachtest: improve multi-region TPC-C tests

This commit changes the multi-region roachtests to target survivability
better, whilst also outputting better metrics. In particular:

* Region survivability now tests shutting down an entire region.
  Workloads now only target one region at a time, with a test for
  cross-region queries.
* Zone survivability now tests shut down one node at a time.
* Added prometheus metrics which can be observed by the snapshot or
  during test invocation.

Release note: None

ajstorm reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> nice!

> Reviewed 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66755#-:-Md2RrdT-HesZhOePAll:b-oxskof)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/cmd/roachtest/tpcc.go, line 459 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66755#-McyEpWl5nrJ6H5r452h:-Md2PDDsDaS0rAmR2B44:b-pkz8ap) ([raw file](https://github.com/cockroachdb/cockroach/blob/d663594e5875825313325bf7836e81d40b526b95/pkg/cmd/roachtest/tpcc.go#L459)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

muy bueno
___
*[pkg/cmd/roachtest/tpcc.go, line 444 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66755#-Md26-8dCBp7jwI6BmG-:-Md26-8dCBp7jwI6BmG0:boj4zoc) ([raw file](https://github.com/cockroachdb/cockroach/blob/f23026f9edef70816d18100069fa168abf7883da/pkg/cmd/roachtest/tpcc.go#L444)):*
> ```Go
> 			zone   string
> 		}{
> 			{region: "us-east1", zone: "us-east1-b"},
> ```

This looks much better - thanks for making the change.

One minor nit: should it be `zones` instead of `zone`?  

Also, one "take-it-or-leave-it-nit": It just took me many minutes to realize that the `region` field here is only being used to determine the _number_ of regions.  You may want to add a comment to that effect, as well as a comment about how one would scale this up to more nodes.  It's not something that a competent engineer couldn't figure out with a bit of digging, but it may save them (or future you) some time.
___
*[pkg/cmd/roachtest/tpcc.go, line 477 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66755#-Md2PWpACJ6m4YxNhXnC:-Md2PWpACJ6m4YxNhXnD:b8gy1s8) ([raw file](https://github.com/cockroachdb/cockroach/blob/f23026f9edef70816d18100069fa168abf7883da/pkg/cmd/roachtest/tpcc.go#L477)):*
> ```Go
> 							ret,
> 							workloadInstance{
> 								nodes:          option.NodeListOption{i + 1}, // 1-indexed
> ```

FWIW, I'm loving these "indexed" comments.
___
*[pkg/cmd/roachtest/tpcc.go, line 499 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66755#-Md2Rk5C45SZlN51Pt96:-Md2Rk5C45SZlN51Pt97:b3ozcqk) ([raw file](https://github.com/cockroachdb/cockroach/blob/f23026f9edef70816d18100069fa168abf7883da/pkg/cmd/roachtest/tpcc.go#L499)):*
> ```Go
> 				workloadInstances: func() []workloadInstance {
> 					// We run two sets of workloads:
> 					// * for each region, have a workload that speaks SQL which only affects data
> ```

Nice comments here too.


<!-- Sent from Reviewable.io -->

otan commented:
thanks!

bors r=ajstorm
